### ISSUE TITLE
Dataset pre-processing with different batch sizes for positive and negative classes

### ISSUE THREAD

**Issue by AnjanaSenanayake** [welch-lab/SquiggleNet#2]
Hi,

My concern is whether it is possible to create torch tensors with different batch sizes for positive and negative classes. The preprocessing script provided in the repository does take only one input argument for batch size. But the problem is in situations where the positive class dataset is significantly larger than the negative class, we have to restrain the batch size to the size of the small dataset.

My exact scenario:

I have covid SP1 dataset and Zymo datasets. SP1 does have about 1.4M reads and Zymo has around 250K reads. Since we have to define the batch size in preprocessing script you have provided, the maximum size we can have is 250K. Though it will include all of the Zymos reads, it will miss 70% of the SP1 reads. Training the model with such a partial and incomplete dataset we can't expect to have the best-trained model.

What I want:

I want to create a dataset with different batch sizes for positive and negative classes. In my exact scenario, something like below.

- positive class Covid SP1 - 1M reads
- negative class Zymo - 200K reads

Hope you may be kindly able to provide some help if the problem is valid and if not some insight on where I have been mistaken.

Thank you.

**Comment by Anbyew** [comment]
Hello! Thanks for bringing up the issue.
Training with imbalanced dataset is a very realistic and yet tough problem. The solution kind of depends on you goal. 

In my personal practice, I would try to balance the training dataset as much as can (the test dataset could be very much imbalanced). Some of the methods including: collecting more data, upsampling/downsampling data, etc. The reason for that is that we want the model to learn the actual difference among the samples instead of learning the induced data distribution bias caused by the imbalance. An extreme example would be if a dataset contains 99% A and 1%B, a model could always predict A regardless of the input and still achieve 99% accuracy, which isn't quite what we want the model to do.

That's being said, there are still ways to combat training with imbalanced data. For one, adopting different evaluation metrics to fit your goal (e.g. instead of accuracies, use precision, recall, F1 score, etc.). You could also use weighted loss function to upscale the underrepresented samples. 

Some of the reference: [8 Tactics to Combat Imbalanced Classes in Your Machine Learning Dataset](https://machinelearningmastery.com/tactics-to-combat-imbalanced-classes-in-your-machine-learning-dataset/),[ Handling imbalanced datasets in machine learning](https://towardsdatascience.com/handling-imbalanced-datasets-in-machine-learning-7a0e84220f28), etc.

Hope this helps!

**Comment by AnjanaSenanayake** [comment]
Hi, Thank you very much for the kind response. Yes, I believe I should consider some different evaluation metrics in my specific case as you have mentioned. Many thanks.
